_episodes/10-diagnostics.md

```diff
@@ -0,0 +1,271 @@
+>    }
+>    return record
+>
+>
+>def compute_diagnostic(filename):
+>    """Compute an example diagnostic."""
+>    logger.debug("Loading %s", filename)
+>    cube = iris.load_cube(filename)
+>
+>    logger.debug("Running example computation")
+>    cube = iris.util.squeeze(cube)
+>    return cube
+>
+>
+>def plot_diagnostic(cube, basename, provenance_record, cfg):
+>    """Create diagnostic data and plot it."""
+>
+>    # Save the data used for the plot
+>    save_data(basename, provenance_record, cfg, cube)
+>
+>    if cfg.get('quickplot'):
+>        # Create the plot
+>        quickplot(cube, **cfg['quickplot'])
+>        # And save the plot
+>        save_figure(basename, provenance_record, cfg)
+>
+>
+>def main(cfg):
+>    """Compute the time average for each input dataset."""
+>    # Get a description of the preprocessed data that we will use as input.
+>    input_data = cfg['input_data'].values()
+>
+>    # Demonstrate use of metadata access convenience functions.
+>    selection = select_metadata(input_data, short_name='tas', project='CMIP5')
+>    logger.info("Example of how to select only CMIP5 temperature data:\n%s",
+>                pformat(selection))
+>
+>    selection = sorted_metadata(selection, sort='dataset')
+>    logger.info("Example of how to sort this selection by dataset:\n%s",
+>                pformat(selection))
+>
+>    grouped_input_data = group_metadata(input_data,
+>                                        'variable_group',
+>                                        sort='dataset')
+>    logger.info(
+>        "Example of how to group and sort input data by variable groups from "
+>        "the recipe:\n%s", pformat(grouped_input_data))
+>
+>    # Example of how to loop over variables/datasets in alphabetical order
+>    groups = group_metadata(input_data, 'variable_group', sort='dataset')
+>    for group_name in groups:
+>        logger.info("Processing variable %s", group_name)
+>        for attributes in groups[group_name]:
+>            logger.info("Processing dataset %s", attributes['dataset'])
+>            input_file = attributes['filename']
+>            cube = compute_diagnostic(input_file)
+>
+>            output_basename = Path(input_file).stem
+>            if group_name != attributes['short_name']:
+>                output_basename = group_name + '_' + output_basename
+>            provenance_record = get_provenance_record(
+>                attributes, ancestor_files=[input_file])
+>            plot_diagnostic(cube, output_basename, provenance_record, cfg)
+>
+>
+>if __name__ == '__main__':
+>
+>    with run_diagnostic() as config:
+>        main(config)
+>
+>~~~
+>{: .language-python}
+>
+{:.solution}
+
+> ## What is the starting point of the diagnostic?
+>
+> Can you spot the main function in the Python code above? How many references 
+>do you see to this function?
+>
+> > ## Answer
+> >
+> > The main function is defined in the middle of this script and is called near the very
```

Comment:
As you probably also noticed this needs a reference to a line number :-)

Reply:
Yes..I am not sure how best to do this. I will complete most of the content and revisit this.